lib/galaxy/config/__init__.py

```diff
@@ -549,6 +549,12 @@ def _process_config(self, kwargs):
         self.database_engine_options = get_database_engine_options(kwargs)
         self.database_create_tables = string_as_bool(kwargs.get('database_create_tables', 'True'))
         self.database_encoding = kwargs.get('database_encoding')  # Create new databases with this encoding
+
+        # Database data tables connection
+        if not self.data_table_connection:  # Provide default if not supplied by user
+            db_path = self._in_data_dir('data_tables.sqlite')
+            self.data_table_connection = 'sqlite:///%s?isolation_level=IMMEDIATE' % db_path
```

Comment:
Can you make this a template?
I can see a use case where data tables need to be stored in a different db per table (sqlite databases)
template values should include table id and tool repo id

Reply:
SQLite is the default, same for the regular Galaxy database. However, for a productive deployment I'd see PostgreSQL (or MySQL) as a better suited system. By making this a template, I'd effectively cause a lot more checks to use it everywhere, thereby greatly increasing the complexity. Whereas, if people want to use it in the most efficient way, they'd probably rely on PostgreSQL or MySQL anyway.   
   
SQLite for me is the default, but not the recommended productive setup. Just like with the regular database connection.

Reply:
The issue of forcing all tables into a single location is it is incompatible with deployments that mount in loc files rather than having a data manager generate them. If we are replacing loc files, the ability to keep the database file with the tool needs to be retained.

Reply:
This system runs alongside the loc files. It doesn't directly replace them.   
Technically what you want could be done by using the file path system.   
For instance a system like:
```xml
<tables>
    <!-- Locations of indexes in the Bowtie mapper format -->
    <table name="bowtie_indexes" comment_char="#">
        <columns>value, dbkey, name, path</columns>
        <file path="tool-data/bowtie_indices.sqlite" />
    </table>
</tables>
```
is what you mean if I am correct. Which indeed isn't possible with this pull request. However most of the code could technically be reused to accomplish this.

Reply:
ah, yes. That would be a reasonable alternative.

Reply:
The problem with that approach is, that especially when working with data managers and tools (and their updates), that Galaxy will have to merge all the SQLite files on run-time to get one coherent data table. As each data manager and each tool using a data table have a `data_table_conf.xml.sample` (or similar) file in their repository or the Tool Shed blocks it (I believe that's the case). So for each data manager or tool update it'd add one SQLite file for the same data table. So for querying you actually may lose some of the benefits of SQL.   
This is what currently happens for the Tabular Tool Data Table (loc files).

Reply:
Yes, I was wondering how you would handle that. Would it be better to remove the ability to split the table across multiple tools and just have a column that tracks the last tool to update the row?
This way when a tool is uninstalled the data for that tool can optionally be deleted from the table?

Reply:
I think I misunderstood your view on the data table system initially. I read replacing the loc files with a larger and replicable database. I didn't consider continuing to keep it on the filesystem. Which brings with it a slew of other problems that I've so far avoided... It may be as simple as adding support for the `<file path="tool-data/bowtie_indices.sqlite" />` elements though in its easiest form, whereby you may not have multiple files that need merging...   
I think that may require an extra PR, because that's a significant change to the data managers if the ability to merge files is lost in this scenario.

Reply:
The issue is it tries to cram all data into one database. I need the ability to choose different database solutions depending on the access patterns and size of data for each table.

Reply:
In that case you'd effectively need a special configuration file just for all your data tables that use databases all with connection strings. While this PR doesn't account for that and provides a general solution, it may also be an option to just store it all in a PostgreSQL database, whether it's big or small. PostgreSQL should be able to handle it fine regardless.